cpp/src/parquet/column_reader.cc

```diff
@@ -333,14 +340,16 @@ void SerializedPageReader::InitDecryption() {
   // Prepare the AAD for quick update later.
   if (crypto_ctx_.data_decryptor != nullptr) {
     ARROW_DCHECK(!crypto_ctx_.data_decryptor->file_aad().empty());
+    data_decryptor_ = std::make_shared<Decryptor>(*crypto_ctx_.data_decryptor);
```

Comment:
So this is making a copy of the `Decryptor` object. But `Decryptor::aes_decryptor_` is a `std::shared_ptr<encryption::AesDecryptor>`, so this is merely going to increment the shared_ptr's reference count, right?

Reply:
I think this means that we want to change:
```c++
std::shared_ptr<Decryptor> GetColumnDataDecryptor(
    const ColumnCryptoMetaData* crypto_metadata, InternalFileDecryptor* file_decryptor);
```
into:
```c++
std::function<std::shared_ptr<Decryptor>()> GetColumnDataDecryptor(
    const ColumnCryptoMetaData* crypto_metadata, InternalFileDecryptor* file_decryptor);
```
so that we can get a new Decryptor instance every time we need one.

(same for the meta decryptor, of course)

Reply:
You are right,

    data_decryptor_ = std::make_shared<Decryptor>(*crypto_ctx_.data_decryptor)

could be simplified to

    data_decryptor_ = crypto_ctx_.data_decryptor

which should increment the reference counter and therefore be equivalent.

Reply:
With `std::function<std::shared_ptr<Decryptor>()> GetColumnDataDecryptor` you are implying `CryptoContext` stores that factory rather than a `std::shared_ptr<Decryptor> data_decryptor`, so we can create a new data decryptor each time we call `InitDecryption`?

Reply:
Yes, exactly!

Reply:
I have applied that change. But something is not right. running test `file_parquet_encryption_test` fails because `GetColumnDecryptor` is now called with null `file_decryptor`.

Reply:
Ah, I'll take a look.

Reply:
Ok, I found the issue. When creating a factory function, we must first compute encryption parameters and store them in the closure that's returned to the caller. I took the liberty of pushing some changes to this PR, I hope you don't mind.